// @bundle-stats/ui/lib-esm/components/metrics-table-search/metrics-table-search.js
import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { Icon } from '../../ui/icon';
import css from './metrics-table-search.module.css';
export var MetricsTableSearch = function MetricsTableSearch(props) {
  var className = props.className,
      placeholder = props.placeholder,
      search = props.search,
      updateSearch = props.updateSearch;
  var rootClassname = cx(css.root, className);
  return /*#__PURE__*/React.createElement("div", {
    className: rootClassname
  }, /*#__PURE__*/React.createElement("input", {
    className: cx('ui-input', 'ui-input--small', css.input),
    placeholder: placeholder,
    onChange: function onChange(event) {
      return updateSearch(event.target.value);
    },
    value: search
  }), search && /*#__PURE__*/React.createElement("button", {
    className: cx('ui-button', 'ui-button--clear', css.cancelButton),
    type: "button",
    onClick: function onClick() {
      return updateSearch('');
    }
  }, /*#__PURE__*/React.createElement(Icon, {
    className: css.cancelButtonIcon,
    glyph: "cancel"
  })));
};
MetricsTableSearch.propTypes = {
  className: PropTypes.string,
  placeholder: PropTypes.string,
  search: PropTypes.string.isRequired,
  updateSearch: PropTypes.func.isRequired
};
MetricsTableSearch.defaultProps = {
  className: '',
  placeholder: ''
};
