// @bundle-stats/ui/lib/components/job-name/job-name.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.JobName = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _tooltip = require("../../ui/tooltip");

var _jobNameModule = _interopRequireDefault(require("./job-name.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var JobNameLabel = function JobNameLabel(_ref) {
  var internalBuildNumber = _ref.internalBuildNumber,
      labelProps = _ref.labelProps;
  return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement("span", labelProps, "Job"), /*#__PURE__*/_react["default"].createElement("span", null, "#".concat(internalBuildNumber)));
};

JobNameLabel.defaultProps = {
  labelProps: {}
};
JobNameLabel.propTypes = {
  internalBuildNumber: _propTypes["default"].number.isRequired,
  labelProps: _propTypes["default"].shape({
    className: _propTypes["default"].string
  })
};

var JobName = function JobName(_ref2) {
  var Component = _ref2.as,
      title = _ref2.title,
      internalBuildNumber = _ref2.internalBuildNumber,
      render = _ref2.render,
      restProps = _objectWithoutProperties(_ref2, ["as", "title", "internalBuildNumber", "render"]);

  return /*#__PURE__*/_react["default"].createElement(_tooltip.Tooltip, _extends({
    className: _jobNameModule["default"].root,
    as: Component,
    title: title
  }, restProps), render({
    internalBuildNumber: internalBuildNumber,
    labelProps: {
      className: _jobNameModule["default"].label
    }
  }));
};

exports.JobName = JobName;
JobName.defaultProps = {
  as: 'span',
  title: '',
  render: JobNameLabel
};
JobName.propTypes = {
  /** Rendered component */
  as: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].func]),

  /** Tooltip title */
  title: _propTypes["default"].string,

  /** Job internal build number */
  internalBuildNumber: _propTypes["default"].number.isRequired,

  /** Render function */
  render: _propTypes["default"].func
};
