// @bundle-stats/utils/lib/config/component-links.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.COMPONENT = exports.SECTIONS = exports.PACKAGE_FILTERS = exports.MODULE_FILTERS = exports.ASSET_FILTERS = exports.MODULE_FILE_TYPE = exports.MODULE_CHUNK = exports.ASSET_FILE_TYPE = exports.ASSET_ENTRY_TYPE = void 0;
const ASSET_ENTRY_TYPE = 'et';
exports.ASSET_ENTRY_TYPE = ASSET_ENTRY_TYPE;
const ASSET_FILE_TYPE = 'ft';
exports.ASSET_FILE_TYPE = ASSET_FILE_TYPE;
const MODULE_CHUNK = 'c';
exports.MODULE_CHUNK = MODULE_CHUNK;
const MODULE_FILE_TYPE = 'mft';
exports.MODULE_FILE_TYPE = MODULE_FILE_TYPE;
const ASSET_FILTERS = {
  ASSET: 'asset',
  CHANGED: 'changed',
  ENTRY: 'entrypoint',
  INITIAL: 'initial',
  CHUNK: 'chunk'
};
exports.ASSET_FILTERS = ASSET_FILTERS;
const MODULE_FILTERS = {
  CHANGED: 'changed'
};
exports.MODULE_FILTERS = MODULE_FILTERS;
const PACKAGE_FILTERS = {
  CHANGED: 'changed',
  DUPLICATE: 'duplicate'
};
exports.PACKAGE_FILTERS = PACKAGE_FILTERS;
const SECTIONS = {
  TOTALS: 'overview',
  ASSETS: 'assets',
  MODULES: 'modules',
  PACKAGES: 'packages'
};
exports.SECTIONS = SECTIONS;
const COMPONENT = {
  BUNDLE_ASSETS: 'ba',
  BUNDLE_MODULES: 'bm',
  BUNDLE_PACKAGES: 'bp'
};
exports.COMPONENT = COMPONENT;
