// @bundle-stats/ui/lib/ui/empty-set/empty-set.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EmptySet = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _emptySetModule = _interopRequireDefault(require("./empty-set.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var EmptySet = function EmptySet(props) {
  var resources = props.resources,
      filtered = props.filtered,
      resetFilters = props.resetFilters;
  var message = filtered ? "No ".concat(resources, " found.") : "No ".concat(resources, " available.");
  return /*#__PURE__*/_react["default"].createElement("div", {
    className: _emptySetModule["default"].root
  }, /*#__PURE__*/_react["default"].createElement("p", {
    className: _emptySetModule["default"].text
  }, message), filtered && /*#__PURE__*/_react["default"].createElement("button", {
    className: _emptySetModule["default"].action,
    type: "button",
    onClick: function onClick() {
      return resetFilters();
    }
  }, "View all"));
};

exports.EmptySet = EmptySet;
EmptySet.propTypes = {
  resources: _propTypes["default"].string.isRequired,
  filtered: _propTypes["default"].bool.isRequired,
  resetFilters: _propTypes["default"].func.isRequired
};
